proposals/lip_define_state_and_state_transitions_of_dpos_module.md

```diff
@@ -0,0 +1,1481 @@
+        * For each element `weightSnapshot` in the `delegateWeightSnapshot` array, there is an element `validator` in the `validators` array with `validator.address == weightSnapshot.delegateAddress`.   
+    * All values of the `genesisData.initDelegates` array must be unique, must have length `ADDRESS_LENGTH`, and must be equal to `validator.address` for `validator` an element of `validators`.
+    * The `genesisData.initDelegates` array must have length less than or equal to `NUMBER_ACTIVE_DELEGATES`.
+
+* For each entry `validator` in `genesisBlockAssetObject.validators`, create an entry in the delegate substore with
+    ```python
+    totalVotesReceived = 0
+    for voter in genesisBlockAssetObject.voters:
+        for sentVote in voter.sentVotes:
+            if sentVote.delegateAddress == validator.address:
+                totalVotesReceived += sentVote.amount
+                if voter.address == validator.address:
+                    selfVotes = sentVote.amount
+            if totalVotesReceived >= 2^64:
+                fail
+  
+    storeKey = validator.address
+    storeValue = {
+        "name": validator.name,
+        "totalVotesReceived": totalVotesReceived,
+        "selfVotes": selfVotes,
+        "lastGeneratedHeight": validator.lastGeneratedHeight,
+        "isBanned": validator.isBanned,
+        "pomHeights": validator.pomHeight,
+        "consecutiveMissedBlocks": validator.consecutiveMissedBlocks
+    } serialized using delegateStoreSchema.
+    ```
+    Further, for every entry `validator` in `genesisBlockAssetObject.validators`, also create an entry in the name substore with 
+    ```python
+    storeKey = validator.name utf8-encoded
+    storeValue = {
+        "delegateAddress": validator.address
+    } serialized using nameStoreSchema.
+    ```
+
+* For each entry `voter` in `genesisBlockAssetObject.voters`, create an entry in the voter substore  with
+    ```python
+    storeKey = uint32be(voter.address)
+    storeValue = {
+        "sentVotes": voter.sentVotes,
+        "pendingUnlocks": voter.pendingUnlocks
+    } serialized using voterStoreSchema.
+    ```
+
+* For each entry `snapshot` in `genesisBlockAssetObject.snapshots`, create an entry in the snapshot substore  with
+    ```python
+    storeKey = uint32be(snapshot.roundNumber)
+    storeValue = {
+        "activeDelegates": snapshot.activeDelegates,
+        "delegateWeightSnapshot": snapshot.delegateWeightSnapshot
+    } serialized using snapshotStoreSchema.
+    ```
+
+* Create an entry in the genesis data substore with
+    ```python
+    storeKey = EMPTY_BYTES
+    storeValue = {
+        "height": block header height of the genesis block,
+        "initRounds": genesisBlockAssetObject.genesisData.initRounds,
+        "initDelegates": genesisBlockAssetObject.genesisData.initDelegates
+    } serialized using genesisDataStoreSchema.
+    ```
```

Comment:
initializing previous timestamp store is missing

Reply:
yes, added.